src/python/pants/backend/native/subsystems/native_toolchain.py

```diff
@@ -63,211 +67,226 @@ def _libc_dev(self):
-  provided_clangpp = original_llvm_cpp_compiler.cpp_compiler
+  provided_clang = original_llvm_cpp_compiler.cpp_compiler
+
+  # These arguments are shared across platforms.
+  llvm_cpp_compiler_args = [
+    '-x', 'c++', '-std=c++11',
+    # These mean we don't use any of the headers from our LLVM distribution.
+    '-nobuiltininc',
+    '-nostdinc++',
+  ]
 
   if platform.normalized_os_name == 'darwin':
     xcode_clang = yield Get(CppCompiler, XCodeCLITools, native_toolchain._xcode_cli_tools)
-    clang_with_xcode_paths = CppCompiler(
-      path_entries=(provided_clangpp.path_entries + xcode_clang.path_entries),
-      exe_filename=provided_clangpp.exe_filename,
-      library_dirs=(provided_clangpp.library_dirs + xcode_clang.library_dirs),
-      include_dirs=(provided_clangpp.include_dirs + xcode_clang.include_dirs))
-    final_llvm_cpp_compiler = LLVMCppCompiler(clang_with_xcode_paths)
+    working_cpp_compiler = CppCompiler(
+      path_entries=(provided_clang.path_entries + xcode_clang.path_entries),
+      exe_filename=provided_clang.exe_filename,
+      library_dirs=(provided_clang.library_dirs + xcode_clang.library_dirs),
+      include_dirs=(provided_clang.include_dirs + xcode_clang.include_dirs),
+      # On OSX, this uses the libc++ (LLVM) C++ standard library implementation. This is
+      # feature-complete for OSX, but not for Linux (see https://libcxx.llvm.org/ for more info).
+      extra_args=(llvm_cpp_compiler_args + xcode_clang.extra_args))
+    linking_library_dirs = []
+    linker_extra_args = []
   else:
     gcc_cpp_compiler = yield Get(GCCCppCompiler, GCC, native_toolchain._gcc)
     provided_gpp = gcc_cpp_compiler.cpp_compiler
-    clang_with_gpp_libs = CppCompiler(
-      path_entries=provided_clangpp.path_entries,
-      exe_filename=provided_clangpp.exe_filename,
-      # We need this version of GLIBCXX to be able to run, unfortunately.
-      library_dirs=(provided_gpp.library_dirs + provided_clangpp.library_dirs),
-      include_dirs=(provided_clangpp.include_dirs + provided_gpp.include_dirs))
-    final_llvm_cpp_compiler = LLVMCppCompiler(clang_with_gpp_libs)
-
-  yield final_llvm_cpp_compiler
-
-
-@rule(GCCCCompiler, [Select(Platform), Select(NativeToolchain)])
-def select_gcc_c_compiler(platform, native_toolchain):
+    working_cpp_compiler = CppCompiler(
+      path_entries=provided_clang.path_entries,
+      exe_filename=provided_clang.exe_filename,
+      # We need g++'s version of the GLIBCXX library to be able to run, unfortunately.
+      library_dirs=(provided_gpp.library_dirs + provided_clang.library_dirs),
+      # NB: we use g++'s headers on Linux, and therefore their C++ standard library.
+      include_dirs=provided_gpp.include_dirs,
+      extra_args=llvm_cpp_compiler_args)
+    linking_library_dirs = provided_gpp.library_dirs + provided_clang.library_dirs
+    # Ensure we use libstdc++, provided by g++, during the linking stage.
+    linker_extra_args=['-stdlib=libstdc++']
```

Comment:
Am I correct in that we are using the provided lib dirs for linking, but not using them for compilation? This is for LLVM-based toolchains only, correct?

Reply:
`linking_library_dirs` aren't used in compilation, only the compiler object is used. This is only for LLVM toolchains -- this `@rule` is producing the single `LLVMCppToolchain`. The `-stdlib` argument is only recognized on clang, see the [libcxx site](https://libcxx.llvm.org/) (the new LLVM C++ standard library that we are turning off with this option, because it is not complete on Linux yet). See the `LinkerMixin` introduced in b2a750b which was added in response to your other review comment, which makes the difference between these directories more clear.